sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionPollerFn.java

```diff
@@ -0,0 +1,146 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.hcatalog;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.io.hcatalog.HCatalogIO.Read;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.DoFn.UnboundedPerElement;
+import org.apache.beam.sdk.transforms.SerializableComparator;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableList;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hive.hcatalog.common.HCatUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Unbounded poller to listen for new partitions. */
+@UnboundedPerElement
+class PartitionPollerFn extends DoFn<Read, HCatRecordReaderFn.PartitionWrapper> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PartitionPollerFn.class);
+  private transient IMetaStoreClient metaStoreClient;
+  private Map<String, String> configProperties;
+  private String database;
+  private String table;
+  final Boolean shouldTreatSourceAsBounded;
+
+  public PartitionPollerFn(
+      final Map<String, String> properties,
+      final String database,
+      final String table,
+      final Boolean isBounded) {
+    this.configProperties = properties;
+    this.database = database;
+    this.table = table;
+    this.shouldTreatSourceAsBounded = isBounded;
+  }
+
+  /** Outputs newer partitions to be processed by the reader. */
+  @ProcessElement
+  @SuppressWarnings("unused")
+  public ProcessContinuation processElement(
+      final ProcessContext c, RestrictionTracker<PartitionRange, Partition> tracker) {
+    final Read readRequest = c.element();
+    final PartitionRange range = tracker.currentRestriction();
+    final ImmutableList<Partition> allAvailablePartitions = range.getPartitions();
+
+    List<Partition> forSort = new ArrayList<>(allAvailablePartitions);
+    Collections.sort(forSort, readRequest.getPartitionComparator());
+
+    int trueStart;
+    if (tracker.currentRestriction().getLastCompletedPartition() != null) {
+      final int indexOfLastCompletedPartition =
+          forSort.indexOf(tracker.currentRestriction().getLastCompletedPartition());
+      trueStart = indexOfLastCompletedPartition + 1;
+    } else {
+      trueStart = 0;
+    }
+
+    for (int i = trueStart; i < forSort.size(); i++) {
+      if (tracker.tryClaim(forSort.get(i))) {
+        c.output(
+            new HCatRecordReaderFn.PartitionWrapper(
```

Comment:
I am not sure that we need the `PartitionWrapper` class. Can't we just add the missing part of the `Read`, the partition as an attribute of the readRequest (with package private access) and then just pass the readRequest (`Read`) object all around?

Reply:
Good point. After the refactor, since everything is now in Read(s), this can go away.